File: scala/src/test/scala/QuickselectSpec.scala
import org.scalatest._
import org.scalatest.prop._
import org.scalacheck.Gen

class QuickselectSpec extends FlatSpec with GeneratorDrivenPropertyChecks with Matchers {
  def testSelect(selector: (Int, Array[Int]) => Option[Int]) {
    it should "return None when passed an empty array" in {
      selector(0, Array.empty) shouldBe None
    }
    it should "return correctly for a small example" in {
      selector(3, Array(7, 10, 4, 3, 20, 15)) shouldBe Some(7)
    }

    val selectorInput =
      for {
        s <- Gen.nonEmptyContainerOf[Set, Int](Gen.choose(0, Int.MaxValue))
        k <- Gen.choose(1, s.size - 1)
      } yield (k, s)

    it should "be the same as taking the (k-1)th element of sort" in {
      forAll(selectorInput) { case (k, xsSet) =>
        whenever (k > 0 && k < xsSet.size) {
          val sorted = xsSet.toSeq.sorted
          val xs: Array[Int] = util.Random.shuffle(sorted).toArray
          selector(k, xs) should be === sorted.lift(k - 1)
        }
      }
    }
  }

  import Quickselect._

  "sortselect" should behave like testSelect(sortselect)
  "quickselect" should behave like testSelect(quickselect)
}

File: haskell/test/QuickselectSpec.hs
{-# LANGUAGE ScopedTypeVariables #-}
module QuickselectSpec (main, spec) where

import           Test.Hspec
import           Test.QuickCheck

import           Quickselect

import           Data.List           (sort)
import qualified Data.Set            as Set
import           Data.Vector.Unboxed (Vector)
import qualified Data.Vector.Unboxed as V

newtype UniqueNum = UniqueNum {getUniques :: Vector Int}
  deriving (Eq, Show, Ord)

instance Arbitrary UniqueNum where
  arbitrary = do
    nums <- arbitrary `suchThat` (not . Set.null)
    nums' <- shuffle (Set.toList nums)
    return . UniqueNum . V.fromList $ nums'

main :: IO ()
main = hspec spec

testSelect f = do
  it "returns Nothing for an empty vector" $
    f 1 V.empty `shouldBe` Nothing
  it "returns Nothing if k is less than 1" $ property $
    \k (UniqueNum xs) -> k < 1 ==> f k xs === Nothing
  it "returns Nothing if k is greater than vector length" $ property $
    \k (UniqueNum xs) -> k > V.length xs ==> f k xs === Nothing
  it "returns the element for a one-element vector" $ property $
    \x -> f 1 (V.singleton x) === Just x
  it "works for a small example" $ do
    let xs = V.fromList [7, 10, 4, 3, 20, 15]
    f 3 xs `shouldBe` Just 7
  it "functions the same as taking the (k-1)th element of sorted list" $ property $
    \(UniqueNum xs) (Positive k) -> k < V.length xs ==>
      f k xs === Just (sort (V.toList xs) !! (k-1))

spec :: Spec
spec = do
  describe "sortselect" $
    testSelect sortselect
  describe "quickselect" $
    testSelect quickselect
